rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/monitoring/views/CompactFlowPanel.java

```diff
@@ -0,0 +1,426 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive_grid_cloud_portal.rm.client.monitoring.views;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import org.ow2.proactive_grid_cloud_portal.rm.client.CompactView;
+import org.ow2.proactive_grid_cloud_portal.rm.client.NodeSource;
+
+import com.google.gwt.user.client.ui.FlowPanel;
+
+
+/**
+ * It is flow panel to draw compact representation of hodesources/hosts/nodes.
+ * Underlying stucture (FlowPanel) is kind of ArrayList.
+ * However, we have hierarchical structure (model)
+ * So to represent model as flat structure we have to be able to identify
+ * index of model elements. Where index is the position of item from model in
+ * FlowPanel. Plus we should be quite fast in finding indexes (in case we have 20K nodes).
+ *
+ * Thus, to have at the same time hierarchical structure (model) and
+ * be able to find any index of this model we keep always updated tiles field
+ * in HierarchyNodeSource and HierarchyHost.
+ * For example: we want to update node. For this we need to find index of the tile which represents
+ * this node. Initially index is zero. Then we sequentially go through nodesources,
+ * hosts, and nodes. Until we find our node, we add to index value of tiles field of HierarchyNodeSource
+ * and HierarchyHost, and 1 for the case of Node.
+ *
+ * Host tiles are ephemeral because CompactFlowPanel handle them by it self.
+ * Add soon there is node but there is no dedicated host for this node, host will be created.
+ * Also, host will be deleted automatically as soon as it does not have nodes anymore.
+ *
+ */
+public class CompactFlowPanel extends FlowPanel {
+
+    protected List<HierarchyNodeSource> model = new LinkedList<>();
+
+    private CompactView.Tile curSelTile;
+
+    public CompactFlowPanel() {
+        super();
+        this.setWidth("100%");
+        // removes the vertical space between lines
+        this.getElement().getStyle().setProperty("lineHeight", "0");
+    }
+
+    public CompactView.Tile getCurSelTile() {
+        return curSelTile;
+    }
+
+    public void setCurSelTile(CompactView.Tile curSelTile) {
+        this.curSelTile = curSelTile;
+    }
+
+    public void drawNodeSource(CompactView.Tile nsTile) {
+        model.add(new HierarchyNodeSource(nsTile.getNodesource()));
+        model.sort(Comparator.comparing(a -> a.getNodeSource().getSourceName()));
```

Comment:
Here it seems we sort a linked list each time we draw a new node source. We just need to make sure that we will only target 20k nodes and not 20k node sources otherwise this will be too costly.

Reply:
We sort only NodeSources, as I understood we should have 20K nodes and not nodeSources